### PR TITLE
Oauth username provider suffix added

### PR DESCRIPTION
**Pull request by asSqr**
I implement the issue https://github.com/go-gitea/gitea/issues/7014. 

In preparation for introducing the second login source, usernames should be namespace-wise. The suffix "-{provider}" will be added automatically.

### CODE REVIEW

**Comment by wxiaoguang** [comment]
1. It breaks existing users (they won't be able to login as their original username)
2. Some users may not want that suffix

**Comment by asSqr** [comment]
Create the provider db table and the namespace field representing the provider in the user model → If some providers are modified, we have only to update the provider table.

This strategy seems good.

**Comment by asSqr** [comment]
I have implemented the above comment. I'm sorry that I have made a trivial mistake and bother you.

File: models/oauth.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	db.RegisterModel(new(OAuth))
+}
+
+func getOAuthByID(e db.Engine, id int64) (*OAuth, error) {
+	o := new(OAuth)
+	has, err := e.ID(id).Get(o)
+	if err != nil {
+		return nil, err
+	} else if !has {
+		return nil, ErrUserNotExist{id, "", 0}
```
Comment (Gusted) [inline models/oauth.go:26]: `ErrUserNotExist` Wrong error to give here.

File: models/oauth.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+type OAuth struct {
+	ID   int64 `xorm:"pk autoincr"`
+	Name string
+}
+
+func init() {
+	db.RegisterModel(new(OAuth))
+}
+
+func getOAuthByID(e db.Engine, id int64) (*OAuth, error) {
```
Comment (Gusted) [inline models/oauth.go:20]: These function is essentially deadcode as it isn't used other than `GetOAuthByID` which on itself isn't used.

File: models/oauth.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+// GetOAuthByName returns oauth by given name.
+func GetOAuthByName(name string) (*OAuth, error) {
+	return getOAuthByName(db.GetEngine(db.DefaultContext), name)
+}
+
+func getOAuthByName(e db.Engine, name string) (*OAuth, error) {
+	if len(name) == 0 {
+		return nil, ErrUserNotExist{0, name, 0}
```
Comment (Gusted) [inline models/oauth.go:43]: Ditto wrong error.

File: models/oauth.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	o := &OAuth{Name: strings.ToLower(name)}
+	has, err := e.Get(o)
+	if err != nil {
+		return nil, err
+	} else if !has {
+		return nil, ErrUserNotExist{0, name, 0}
```
Comment (Gusted) [inline models/oauth.go:50]: Ditto wrong error.

File: models/oauth.go (hunk 2 above)
Comment (Gusted) [reply models/oauth.go:20]: General note, a lot of dead code functions are in these files...

File: models/user.go [hunk 5]
```diff
@@ -162,6 +162,9 @@ type User struct {
 	DiffViewStyle       string `xorm:"NOT NULL DEFAULT ''"`
 	Theme               string `xorm:"NOT NULL DEFAULT ''"`
 	KeepActivityPrivate bool   `xorm:"NOT NULL DEFAULT false"`
+
+	// OAuth
+	OAuthProvider       int64
```
Comment (Gusted) [inline models/user.go:167]:
```suggestion
	OAuthProviderID       int64
```

You will also need a migration for this - as existing users won't have any.

File: routers/web/user/auth.go [hunk 6]
```diff
@@ -649,14 +649,28 @@ func SignInOAuthCallback(ctx *context.Context) {
 				ctx.ServerError("CreateUser", err)
 				return
 			}
+
+			oauthProviderName := strings.ToLower(provider)
```
Comment (Gusted) [inline routers/web/user/auth.go:653]: What does this actually? 

This adds the name of the oauth into a new database table, link within the user to the ID of that oauth. But that doesn't solve the issue of adding a suffix.

Also, we already have this, look at the `LoginSource` field ;) which should give you the same way of getting the Oauth's name.

File: models/oauth.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
+package models
```
Comment (lunny) [inline models/oauth.go:1]: copyright head missing

File: models/oauth.go (hunk 2 above)
Comment (lunny) [reply models/oauth.go:20]: I think we already have the table, why should we create a new one?

File: models/user.go (hunk 5 above)
Comment (lunny) [reply models/user.go:167]: We have an `external_account` table, I think it's already ready.

**Review by lunny — CHANGES_REQUESTED** [review]
I think most of them have been implement

**Comment by yardenshoham** [comment]
This is in progress for a while now, I'm closing it to avoid stale pull requests. Please reopen when it's ready for review.

**Pull request closed without merging**